re_data_ui/src/components/FlowGraph.tsx

```diff
@@ -0,0 +1,202 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import React, {
+  MouseEvent as ReactMouseEvent, ReactElement, useCallback,
+  useEffect, useRef, useState,
+} from 'react';
+import ReactFlow, {
+  ConnectionLineType, Controls, Edge, Elements, getIncomers, getOutgoers,
+  isEdge, isNode, Node, OnLoadParams, ReactFlowProvider,
+  removeElements,
+} from 'react-flow-renderer';
+import { useSearchParams } from 'react-router-dom';
+import '../graph.css';
+import { getLayoutElements } from '../utils';
+import CustomNode from './CustomNode';
+
+export interface Props {
+  data: Elements;
+  disableClick?: boolean;
+}
+
+const nodeTypes = {
+  'custom-node': CustomNode,
+};
+
+function FlowGraph(params: Props): ReactElement {
+  const { data, disableClick = false } = params;
+  const instanceRef = useRef<OnLoadParams | null>(null);
+  const [, setURLSearchParams] = useSearchParams();
+
+  const res = getLayoutElements(data);
+  const [elements, setElements] = useState<Elements>(res);
+
+  useEffect(() => {
+    setElements(res);
+  }, [data]);
+
+  const onElementsRemove = (elementsToRemove: Elements) => {
+    setElements((els) => removeElements(elementsToRemove, els));
+  };
+
+  const getAllIncomer = (node: Node, el: Elements): Node[] => getIncomers(node, el).reduce(
+    (memo: any, incomer) => [
+      ...memo,
+      incomer,
+      ...getAllIncomer(incomer, elements),
+    ],
+    [],
+  );
+
+  const getAllOutgoer = (node: Node, el: Elements): Node[] => getOutgoers(node, el).reduce(
+    (memo: any, outgoer) => [
+      ...memo,
+      outgoer,
+      ...getAllOutgoer(outgoer, el),
+    ],
+    [],
+  );
+
+  const removeHighlightPath = (): void => {
+    const values = elements?.map((elem) => {
+      const element = elem;
+      if (isNode(element)) {
+        element.style = {
+          ...element.style,
+          opacity: 1,
+        };
+      }
+      if (isEdge(element)) {
+        element.animated = false;
+      }
+      return element;
+    });
+
+    setElements(values);
+  };
+
+  const highlightPath = (node: Node, selection: boolean): void => {
+    if (node && elements) {
+      const allIncomer: Node[] = getAllIncomer(node, elements);
+      const allOutgoer: Node[] = getAllOutgoer(node, elements);
+
+      setElements((prevElements) => prevElements?.map((elem) => {
+        const element = elem;
+        const incomerIds = allIncomer.map((i) => i.id);
+        const outgoerIds = allOutgoer.map((o) => o.id);
+
+        if (isNode(element)) {
+          const highlight = element.id === node.id
+              || incomerIds.includes(element.id)
+            || outgoerIds.includes(element.id);
```

Comment:
A set would be more efficient here when there are a large amount of nodes.

Reply:
nice